ts/private/ts_project_worker.js

```diff
@@ -419,37 +498,63 @@ function createProgram(args, initialInputs) {
         noop,
         noop
     );
+    delete host.setTimeout;
+    delete host.clearTimeout;
 
-    host.invalidate = invalidate;
-    host.applyChanges = applyChanges;
-    host.debuglog = debuglog;
 
-    debuglog(`tsconfig: ${tsconfig}`);
-    debuglog(`execroot ${execRoot}`);
+    /** @type {ts.FormatDiagnosticsHost} */
+    const formatDiagnosticHost = {
+        getCanonicalFileName: (path) => path,
+        getCurrentDirectory: sys.getCurrentDirectory,
+        getNewLine: () => sys.newLine,
+    };
+
+    debug(`tsconfig: ${tsconfig}`);
+    debug(`execroot: ${execRoot}`);
 
     const program = ts.createWatchProgram(host);
 
-    return { host, program, checkAndApplyArgs, enablePerformanceAndTracingIfNeeded };
+    return { host, program, checkAndApplyArgs, enablePerformanceAndTracingIfNeeded, setOutput, formatDiagnostics, flushWatchEvents, invalidate };
+
+    function formatDiagnostics(diagnostics) {
+        return `\n${ts.formatDiagnostics(diagnostics, formatDiagnosticHost)}\n`
+    }
+
+    function setOutput(newOutput) {
+        output = newOutput;
+    }
 
-    function setTimeout(cb) {
-        // NB: tsc will never clearTimeout if the index is 0 hence timeout must be truthy. :|
-        // https://github.com/microsoft/TypeScript/blob/d0bfd8caed521bfd24fc44960d9936a891744bb7/src/compiler/watchPublic.ts#L681
-        return taskQueue.push(cb);
+    function write(chunk) {
+        output.write(chunk);
     }
- 
-    function clearTimeout(i) {
-        delete taskQueue[i - 1];
+
+    function exit(exitCode) {
+        debug(`program wanted to exit prematurely with code ${exitCode}`);
     }
 
-    function applyChanges() {
-        for (let i = 0; i < taskQueue.length; i++) {
-            const task = taskQueue[i];
-            delete taskQueue[i]
-            if (task) {
-                task();
-            }  
+    function flushWatchEvents() {
+        for (const [callback, ...args] of watchEventQueue) {
+            callback(...args);
+        }
+        watchEventQueue.length = 0;
+    }
+
+    function invalidate(filePath, kind) {
+        if (filePath.endsWith('.params')) return;
+        debug(`invalidate ${filePath} : ${ts.FileWatcherEventKind[kind]}`);
+        if (kind === ts.FileWatcherEventKind.Deleted) {
+            filesystemTree.remove(filePath);
+        } else if (kind === ts.FileWatcherEventKind.Created) {
+            filesystemTree.add(filePath);
+        } else {
+            filesystemTree.update(filePath);
+        }
+        if (filePath.indexOf("node_modules") != -1 && filesystemTree.isSymlink(filePath) && kind === ts.FileWatcherEventKind.Created) {
```

Comment:
this is the second biggest culprit and is the reason why adding dependencies wouldn't discard some diagnostics;

the weirdest case I have seen;

> this is case 11 in our test.sh but wasn't failing due to previously written state;

image src.ts and BUILD that looks like this 

```ts
import {} from "@nestjs/core" 
```

```starlark
ts_project(
  srcs = ["src.ts"],
 deps = [":node_modules/@nestjs/core"]
)
```

ts_project will run into a bad state if the dependency is missing from the deps initially in a cold build case;
However,  if the dependency was listed on the deps initially and built, then tsc won't run into a bad state and report changes correctly for this dependency. 
this is the reason why adding missing dependencies didn't discard the diagnostics. 

How does this happen you ask? simple

1- tsc tries to look for node_modules finds that there is one at `bazel-bin/node_modules`
2- looks for `bazel-bin/node_modules/@nest/core` to resolve '@nest/core'
3- fails to resolve because it's missing from the deps
4- creates a bunch of paths called failedLookups that looks like
```
bazel-bin/node_modules/@nest/core/index.d.ts
bazel-bin/node_modules/@nest/core.d.ts
bazel-bin/node_modules/@types/node_nestjs_core.d.ts
```
5- fails to discard these entries (hence the diagnostics) because we never report anything as created|changed under that directory because it's a symlink. just the directory itself reported but that doesn't discard any of the cache entries above.